### PR TITLE
[Merged by Bors] - chore: forward-port leanprover-community/mathlib#18949

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/18949

[`order.hom.lattice`@`9d684a893c52e1d6692a504a118bfccbae04feeb`..`7581030920af3dcb241d1df0e36f6ec8289dd6be`](https://leanprover-community.github.io/mathlib-port-status/file/order/hom/lattice?range=9d684a893c52e1d6692a504a118bfccbae04feeb..7581030920af3dcb241d1df0e36f6ec8289dd6be)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Hom/Lattice.lean [hunk 1]
```diff
@@ -1588,3 +1588,263 @@ theorem symm_dual_comp (g : BoundedLatticeHom βᵒᵈ γᵒᵈ) (f : BoundedLat
...
+
+/-! ### `with_top`, `with_bot` -/
+
+namespace SupHom
+variable [SemilatticeSup α] [SemilatticeSup β] [SemilatticeSup γ]
+
+/-- Adjoins a `⊤` to the domain and codomain of a `SupHom`. -/
+@[simps]
+protected def withTop (f : SupHom α β) : SupHom (WithTop α) (WithTop β) where
+  toFun := Option.map f
```
Comment (eric-wieser) [inline Mathlib/Order/Hom/Lattice.lean:1600]: This should be `WithTop.map` too (and in Lean 3)

**Comment by eric-wieser** [comment]
> feat: Free functor Lat ⥤ BddLat

I don't think this is a good name for the PR. This PR does _not_ add the free functor it claims to, as that resides in an unported file and will appear later.

File: Mathlib/Order/Hom/Lattice.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Order/Hom/Lattice.lean:1600]:
```suggestion
  toFun := WithTop.map f
```

Comment (YaelDillies) [reply Mathlib/Order/Hom/Lattice.lean:1600]: Ah, right!

**Comment by joneugster** [comment]
Since this has been around for a while and I'd now like to start porting `BddLat` I tried my best applying @eric-wieser 's suggestions. Does that look good?

File: Mathlib/Order/Hom/Lattice.lean (hunk 1 above)
Comment (joneugster) [reply Mathlib/Order/Hom/Lattice.lean]: done the latter

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**